### PR TITLE
refactor: split select menu types into base and concrete type

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

This splits the select menu component/item types into a base type (`BaseSelectMenu`, `ui.BaseSelect`) and a concrete type (`SelectMenu`, `ui.Select`; these already exist) respectively, in preparation for the new select menu types.

This was cherry-picked from the feature branch to hopefully simplify reviews and to avoid a 2k+ line diff ~~(it's just 1.5k now \o/)~~.

Essentially no functional changes, and all the existing types are preserved in a backwards-compatible manner (i.e. `disnake.ui.Select` still works fine).

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -36,6 +36,8 @@ ignore =
     S110,
     # ignore multiple `##` starting a comment
     E266,
+    # ignore "unused import" errors, already handled by pyright
+    F401
```
Comment (onerandomusername) [inline .flake8:40]: I'd rather we don't make this change, and at least not in this pull. Other things are covered by both flake8 and pyright, and we've been known to find pyright bugs, so this should stay enabled.

File: disnake/components.py [hunk 2]
```diff
@@ -210,18 +213,17 @@ def to_dict(self) -> ButtonComponentPayload:
...
+class BaseSelectMenu(Component):
+    """Represents an abstract select menu from the Discord Bot UI Kit.
 
     A select menu is functionally the same as a dropdown, however
     on mobile it renders a bit differently.
 
-    .. note::
+    The currently supported select menus are:
 
-        The user constructible and usable type to create a select menu is
```
Comment (onerandomusername) [inline disnake/components.py:221]: this line should not be deleted, but modified.

File: disnake/components.py [hunk 3]
```diff
@@ -279,8 +276,47 @@ def to_dict(self) -> SelectMenuPayload:
         return payload
 
 
+class SelectMenu(BaseSelectMenu):
+    """Represents a string select menu from the Discord Bot UI Kit.
+
+    .. versionadded:: 2.0
```
Comment (onerandomusername) [inline disnake/components.py:282]: this is missing the user-constructable line that previously existed.

File: disnake/interactions/message.py [hunk 4]
```diff
@@ -84,7 +84,7 @@ def values(self) -> Optional[List[str]]:
 
     @cached_slot_property("_cs_component")
     def component(self) -> MessageComponent:
-        """Union[:class:`Button`, :class:`SelectMenu`]: The component the user interacted with"""
+        """Union[:class:`Button`, :class:`BaseSelectMenu`]: The component the user interacted with"""
```
Comment (onerandomusername) [inline disnake/interactions/message.py:87]: versionchanged

File: disnake/ui/item.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ class WrappedComponent(ABC):
     The following classes implement this ABC:
 
     - :class:`disnake.ui.Button`
-    - :class:`disnake.ui.Select`
+    - subtypes of :class:`disnake.ui.BaseSelect` (:class:`~disnake.ui.Select`)
```
Comment (onerandomusername) [inline disnake/ui/item.py:45]: this should display `ui.Select` *at minimum* and perhaps the full name.

File: disnake/ui/item.py [hunk 6]
```diff
@@ -81,7 +81,7 @@ class Item(WrappedComponent, Generic[V_co]):
     The current UI items supported are:
 
     - :class:`disnake.ui.Button`
-    - :class:`disnake.ui.Select`
+    - subtypes of :class:`disnake.ui.BaseSelect` (:class:`~disnake.ui.Select`)
```
Comment (onerandomusername) [inline disnake/ui/item.py:84]: this should display `ui.Select` *at minimum* and perhaps the full name.

File: disnake/ui/select/__init__.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+~~~~~~~~~~~~~~~~~~
+
+Select Menu UI Kit Types
+
+:copyright: (c) 2021-present Disnake Development
+:license: MIT, see LICENSE for more details.
+
+"""
+from .base import *
+from .string import *
```
Comment (onerandomusername) [inline disnake/ui/select/__init__.py:14]: also needs to define `__all__` and extend it, like other `__init__`'s

File: .flake8 (hunk 1 above)
Comment (shiftinv) [reply .flake8:40]: This is due to a flake8 bug <sup>:^)</sup>, the alternative would be adding several inline `noqa`s, which wouldn't be great either.

```py
from datetime import timedelta
from typing import Generic, TypeVar

T = TypeVar("T")


class A(Generic[T]):
    ...


class B(A["timedelta"]):
    ...
```

```py
$ flake8 test.py
test.py:1:1: F401 'datetime.timedelta' imported but unused
```

File: disnake/interactions/message.py (hunk 4 above)
Comment (shiftinv) [reply disnake/interactions/message.py:87]: This is a documentation-only change, the `MessageComponent` return type is more concrete and unchanged compared to before. I didn't really feel like listing all select types here (once the remaining 4 get added).

File: disnake/components.py (hunk 2 above)
Comment (shiftinv) [reply disnake/components.py:221]: should be resolved as of https://github.com/DisnakeDev/disnake/pull/800/commits/46c30453d0d22f6c4c4a2ce5eb481dcea9ba9941, if I understood correctly

File: disnake/components.py (hunk 3 above)
Comment (shiftinv) [reply disnake/components.py]: https://github.com/DisnakeDev/disnake/pull/800/commits/46c30453d0d22f6c4c4a2ce5eb481dcea9ba9941

File: disnake/ui/item.py (hunk 5 above)
Comment (shiftinv) [reply disnake/ui/item.py]: https://github.com/DisnakeDev/disnake/pull/800/commits/a6694525818becff893c2fb0909435f6f10fe863

File: disnake/ui/item.py (hunk 6 above)
Comment (shiftinv) [reply disnake/ui/item.py]: https://github.com/DisnakeDev/disnake/pull/800/commits/a6694525818becff893c2fb0909435f6f10fe863

File: disnake/ui/select/__init__.py (hunk 7 above)
Comment (shiftinv) [reply disnake/ui/select/__init__.py:15]: good catch, fixed in https://github.com/DisnakeDev/disnake/pull/800/commits/9357233e48f5f7cf2e1dbb0874bdb54464913482

**Review by onerandomusername — APPROVED** [review]
lgtm

**Pull request merged**